### PR TITLE
Dry run implementation

### PR DESCRIPTION
**Pull request by hal-henningsmoen**
Review :white_check_mark: Merge :no_entry_sign: 
Still developing.
Reviewers, please give the compose_job_file method a careful look because this is where my change actually affects a standard, as opposed to dry run, processing path.
@dhallman please review.
This uses Dean's earlier work to add dry-run functionality to Hodor so .properties and .properties.erb files are generated and deposited in the appropriate .tmp directory but file is not deployed to server and job is not run.
Wanted this out before further work is done so we can generate a set of _before_ .properties files for comparison with _after_ files generated when new system is in place.

### CODE REVIEW

File: lib/tasks/oozie.thor [hunk 1]
```diff
@@ -252,8 +252,10 @@ module Hodor
        explanation, view the "jobs.yml" topic, as follows:
        \x5   $ hodor oozie:topic jobs.yml
       ].unindent(8)
+      method_option :dry_run, type: :boolean, aliases: "-d", default: false,
+        desc: "Just display the computed job.properties file. Don't actually deploy or run the job."
```
Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: Suggest: "Display computed job.properties file without running or deploying associated job."

Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: Good idea.

Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: Actually should be:
"Generate computed job.properties file without running or deploying associated job."

Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: Yes, good, especially since this is now opensource, and we need to be perfectly clear in our descriptive language, especially for those whose first language is not English. Question: does this "generate and display" or "generate and write [somewhere else]" the computed job.properties file?

Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: BTW, shouldn't this desc have a first argument like "dry_run [true|false]"? See the desc for 'kill_job' in line 261 for an example: there are two arguments to desc there, first for usage, second for description.

Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: Right now it generates and writes.  I will restructure code so exact location where this is written is displayed if this is a dry run.  Message would be something like "DRY RUN. Generated properties available for inspection at #{full_file_path}"

Will modify desc so consistent with kill_job.  

Also adding another argument to allow a file_name_prefix defaulting to empty string.  This will  simplify identifying the *before* files.

Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: Cool. BTW, where are the specs which cover this new behavior?

Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: Specifically, shouldn't there be coverage in the rspec directory?

Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: I believe the specification for dry run is suggested in the dev plan for DTM-421.  @dhallman had this work on a branch that also included initial work on the new properties generation code.   We decided to get this functionality out quickly in a separate branch so we could generate the full set of .properties files we currently employ for jobs before making other changes to the code. 

Intent was to get this out quickly then add coverage for all new functionality.  I can implement rspec coverage as part of this work.  Likely can still have it ready to merge by EOD.

Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: Sorry, by "spec" I did mean "rspec," not "specification." You said, "Intent was to get this out quickly then add coverage for all new functionality. I can implement rspec coverage as part of this work." -- *All* work should be done test-first. Let's discuss this offline.

**Comment by russellwhitaker** [comment]
I have added a .travis.yml file to this PR and enabled the web service associated with it @hal-henningsmoen @dhallman

**Comment by dhallman** [comment]
Oh, great. Thanks @russellwhitaker.

**Comment by dhallman** [comment]
The anything argument is defined here: https://www.relishapp.com/rspec/rspec-mocks/v/2-7/docs/argument-matchers/general-matchers

**Comment by russellwhitaker** [comment]
@dhallman interesting, thanks: fascinating that RSpec allows the use of the bare identifier 'anything' in certain positions but seems to require quoted "anything" in other positions.

File: lib/tasks/oozie.thor [hunk 2]
```diff
@@ -248,12 +248,19 @@ module Hodor
       long_desc %Q[
        The run_job command reads the contents of the jobs.yml file located
        in your current directory, composes a job.properties file and submits
-       the indicated driver workflow for execution by Oozie. For a fuller
-       explanation, view the "jobs.yml" topic, as follows:
+       the indicated driver workflow for execution by Oozie. If a job.properties
+       or job.properties.erb file is provided this will be used for property derivation.
+       If none is provided properties will be derived by composing property files in
+       the directory structure from the top down. For a fuller explanation, view the "jobs.yml" topic, as follows:
```
Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: Documentation of option to pass in a job.properties(.erb) file.  Improvements to wording appreciated @russellwhitaker @dhallman

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+language: ruby
```
Comment (russellwhitaker) [inline .travis.yml]: Note that this file - which I myself added - hasn't yet been tested, because it can really only be tested by - ta da! - a merge to master, so we'll see. :)

File: lib/hodor/api/oozie.rb [hunk 4]
```diff
@@ -147,27 +147,44 @@ def compose_job_file(direct_job = nil)
             result
           }
           FileUtils.mkdir './.tmp' unless Dir.exists?('./.tmp')
-          propfile = File.expand_path(".tmp/runjob.properties.erb", pwd)
-          File.open(propfile, "w") do |f|
+          composite_properties_file = File.expand_path(".tmp/runjob.properties.erb", pwd)
```
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: This identifier seems to express your original intention best, yes.

File: lib/hodor/api/oozie.rb [hunk 5]
```diff
@@ -147,27 +147,44 @@ def compose_job_file(direct_job = nil)
...
+          File.open(composite_properties_file, "w") do |f|
             f.puts composite_jobfile
           end
-          dest_file = propfile.sub(/\.erb$/,'')
-          erb_expanded = env.erb_load(propfile)
-          File.open(dest_file, 'w') { |f| f.write(erb_expanded) }
-          dest_file
+
+          out_file = composite_properties_file.sub(/\.erb$/,'')
```
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: Lines 154,155: what's the distinction (to the reader looking at these identifiers later in the code) between "out_file" and "dest_file"?

File: lib/hodor/api/oozie.rb [hunk 6]
```diff
@@ -147,27 +147,44 @@ def compose_job_file(direct_job = nil)
...
+          dest_file = generate_and_write_job_file(out_file, composite_properties_file, prefix)
         else
-          raise "Job file '#{direct_job}' not found" if !File.exists?(direct_job)
+          raise "Job file '#{direct_job}' not found" unless File.exists?(direct_job)
```
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: Ah, a bug fix on this line. :)

File: .travis.yml (hunk 3 above)
Comment (hal-henningsmoen) [inline .travis.yml]: Isn't that what we see at the bottom of this PR page?  "All checks have passed..."

File: lib/hodor/api/oozie.rb [hunk 7]
```diff
@@ -147,27 +147,44 @@ def compose_job_file(direct_job = nil)
...
-            erb_expanded = env.erb_load(direct_job)
-            dest_file = "/tmp/#{File.basename(direct_job.sub(/.erb$/,''))}"
-            File.open(dest_file, 'w') { |f| f.write(erb_expanded) }
+            out_file = "/tmp/#{File.basename(direct_job.sub(/.erb$/,''))}"
+            dest_file = generate_and_write_job_file(out_file, direct_job, prefix)
           else
             dest_file = direct_job
+            out_file = append_prefix_to_filename(destfile, prefix)
+            unless prefix.blank?
+              FileUtils.cp(dest_file, out_file)
```
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: Line 166: cf. my earlier comment on lines 154,155, here's a place where confusion can happen. You and I both know a filesystem copy command in any language usually takes the form of something like "copy(in, out)", so this ends up looking a bit like "copy(some_out, some_other_out)".

File: lib/hodor/api/oozie.rb (hunk 6 above)
Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: Style, I find if !  less clear than unless.

File: lib/hodor/api/oozie.rb [hunk 8]
```diff
@@ -147,27 +147,44 @@ def compose_job_file(direct_job = nil)
...
+              dest_file = out_file
+            end
           end
         end
         dest_file
       end
 
+      def generate_and_write_job_file(file_name, in_file, prefix='')
+        erb_expanded = env.erb_load(in_file)
```
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: Lines 175,176: inconsistent spaces absent/present around "=". Spaces around both. 176, be like 175.

File: lib/hodor/api/oozie.rb [hunk 9]
```diff
@@ -147,27 +147,44 @@ def compose_job_file(direct_job = nil)
...
+        out_file=append_prefix_to_filename(file_name, prefix)
+        File.open(out_file, 'w') { |f| f.write(erb_expanded) }
+        out_file
+      end
+
+      def append_prefix_to_filename(file_name, prefix='')
+        out_file = String.new(file_name)
+        out_file.insert(-(1 + out_file.reverse.index(/[\/]/)), prefix)
```
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: Fascinating... trying to wrap my head around the intention of this. Seems clever. Should it be less clever? Also, I'm pretty sure there's a one-line solution alternative to lines 182-184.

File: lib/hodor/api/oozie.rb [hunk 10]
```diff
@@ -180,18 +197,19 @@ def deploy_job(job, clean_deploy)
         end
       end
 
-      def run_job(job = nil)
+      def run_job(job = nil, dry_run = false, file_prefix='')
         if job && (job =~ /job.properties.erb$/ || job =~ /job.properties/)
           # Direct mode: dispatch job.properties file supplied explicitly
```
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: Avoid inline explanatory comments like this. Either omit comments entirely, and obviate the need for explanation with common idioms and good identifier naming, or place any "can't otherwise explain my rationale for doing this exceptional thing" comment before the body of the definition.

File: lib/hodor/api/oozie.rb (hunk 7 above)
Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: Agreed, I can find a cleaner way to do this messy thing.   In essence this is here so prefix and dry_run options work with case where no file is actually generated.  That being said I have latitude with identifier naming that will make this clearer.

File: lib/tasks/oozie.thor [hunk 11]
```diff
@@ -248,12 +248,19 @@ module Hodor
...
        \x5   $ hodor oozie:topic jobs.yml
       ].unindent(8)
+      method_option :dry_run, type: :boolean, aliases: "-d", default: false,
+        desc: "Generate computed job.properties file without running or deploying associated job."
+      method_option :file_name_prefix, type: :string, aliases: '-p', default: '',
+                    desc: "Add a prefix to job properties filename.This is primarily for use with dryrun"
       def run_job(*args)
-        oozie.run_job(args.length > 0 ? args[0] : nil)
+        outfile = oozie.run_job(args.length > 0 ? args[0] : nil, options[:dry_run], options[:file_name_prefix])
+        puts "Dry run: generated properties available for inspection at #{outfile}"  if options[:dry_run]
```
Comment (dhallman) [inline lib/tasks/oozie.thor]: Avoid using "puts". Use "logger.info" instead. Otherwise, specs are littered with puts output.

Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: Had to step away, hadn't gotten down to that section yet myself. Yes, what @dhallman says: use logger.info for this; see what Dean does elsewhere in the repo, where he defines logging levels in one file and uses it in 5 others (4 .thor, 1 .rb). Besides being a better thing to do, it's also consistent with the existing codebase.

File: .travis.yml (hunk 3 above)
Comment (russellwhitaker) [inline .travis.yml]: Ah, cool! OK, yes, it does do that on branches too. Cool.

File: lib/hodor/api/oozie.rb (hunk 6 above)
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: I need my eyes checked once again. Yes, and yes indeed.

File: lib/hodor/api/oozie.rb (hunk 9 above)
Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: Will simplify.  Most string substitutions modify input string insert actually behaves as insert!

File: lib/hodor/api/oozie.rb [hunk 12]
```diff
@@ -180,18 +197,19 @@ def deploy_job(job, clean_deploy)
...
-          jobfile = compose_job_file
+          jobfile = compose_job_file(nil, file_prefix)
+        end
+        unless dry_run
+          runfile = env.deploy_tmp_file(jobfile)
+          env.ssh "oozie job -oozie :oozie_url -config #{runfile} -run", echo: true, echo_cmd: true
         end
-        runfile = env.deploy_tmp_file(jobfile)
-        env.ssh "oozie job -oozie :oozie_url -config #{runfile} -run", echo: true, echo_cmd: true
+        jobfile if dry_run
```
Comment (dhallman) [inline lib/hodor/api/oozie.rb]: I would suggest returning jobfile unconditionally. The caller applies the same dry_run option check again anyway.

File: lib/hodor/api/oozie.rb (hunk 9 above)
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: More generally, just about any time I see a block where 1.) an object "a" is created and initialized; 2.) object "a" is mutated in place; and 3.) object "a" is returned, I see an easy refactoring to a one-liner, especially in a friendly language like ruby where the output of the last evaluated expression is the return value by default (unless otherwise consumed.)

Also, what necessitates explicitly constructing a new string here? Aren't you already expecting a string as input?

Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: I don't want the side effect of modifying the input string.   Working on a simpler one liner.

Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: In other words:
```ruby
→ irb
2.1.7 :001 > def append_prefix_to_filename(file_name, prefix='')
2.1.7 :002?>   file_name.insert(-(1 + file_name.reverse.index(/[\/]/)), prefix)
2.1.7 :003?>   end
 => :append_prefix_to_filename
2.1.7 :004 > append_prefix_to_filename("mydir/myfile", "myprefix-")
 => "mydir/myprefix-myfile"
2.1.7 :005 >
```

Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: There's your one-liner. You're not "modifying the input string," you're modifying a copy scoped to the block only.

Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: BTW, does your test coverage for this include the edge condition of "zero embedded slashes"?

Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: in irb:
`2.2-head :015 > pwd = FileUtils.pwd
 => "/Users/halhenningsmoen/Documents/ulive/datum/hodor" 
2.2-head :016 > def append_prefix_to_filename(file_name, prefix='')
2.2-head :017?>   file_name.insert(-(1 + file_name.reverse.index(/[\/]/)), prefix)
2.2-head :018?>    end
 => :append_prefix_to_filename 
2.2-head :019 > append_prefix_to_filename(pwd, 'prefix_')
 => "/Users/halhenningsmoen/Documents/ulive/datum/prefix_hodor" 
2.2-head :020 > pwd
 => "/Users/halhenningsmoen/Documents/ulive/datum/prefix_hodor" 
2.2-head :021 > `

Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: Having difficulty reading that last comment @hal-henningsmoen but it looks like you converged on the same alternative.

Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: @russellwhitaker edited output should be clearer that method changes value of input string.

File: lib/tasks/oozie.thor [hunk 13]
```diff
@@ -248,12 +248,19 @@ module Hodor
...
        The run_job command reads the contents of the jobs.yml file located
        in your current directory, composes a job.properties file and submits
-       the indicated driver workflow for execution by Oozie. For a fuller
-       explanation, view the "jobs.yml" topic, as follows:
+       the indicated driver workflow for execution by Oozie. If a job.properties
+       or job.properties.erb file is provided this will be used for property derivation.
+       If none is provided properties will be derived by composing property files in
+       the directory structure from the top down. For a fuller explanation, view the "jobs.yml" topic, as follows:
        \x5   $ hodor oozie:topic jobs.yml
       ].unindent(8)
```
Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: I know this isn't part of the PR, but I'm wondering about this hardcoded argument to "unindent()."

**Comment by russellwhitaker** [comment]
In expanding one of the files in this PR, I found a misspelling that's being carried over without modification in this PR: https://github.com/data-wranglers/hodor/blob/hh/dtm-421/implement_dry_run_property_generation/lib/tasks/oozie.thor#L227

File: lib/hodor/api/oozie.rb (hunk 9 above)
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: @hal-henningsmoen yep, more readable. Yep, we converged on the same solution.

File: lib/tasks/oozie.thor [hunk 14]
```diff
@@ -248,12 +248,19 @@ module Hodor
       long_desc %Q[
        The run_job command reads the contents of the jobs.yml file located
        in your current directory, composes a job.properties file and submits
-       the indicated driver workflow for execution by Oozie. For a fuller
-       explanation, view the "jobs.yml" topic, as follows:
+       the indicated driver workflow for execution by Oozie. If a job.properties
+       or job.properties.erb file is provided this will be used for property derivation.
```
Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: 1.) "...is provided, that file..."
2.) "...it will be used to interpolate property values."

File: lib/tasks/oozie.thor [hunk 15]
```diff
@@ -248,12 +248,19 @@ module Hodor
...
+       If none is provided properties will be derived by composing property files in
```
Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: 1.) "If a file is not provided, properties..."
2.) "...will be derived from values in property files..."

File: lib/tasks/oozie.thor [hunk 16]
```diff
@@ -248,12 +248,19 @@ module Hodor
...
+       the directory structure from the top down. For a fuller explanation, view the "jobs.yml" topic, as follows:
        \x5   $ hodor oozie:topic jobs.yml
       ].unindent(8)
+      method_option :dry_run, type: :boolean, aliases: "-d", default: false,
+        desc: "Generate computed job.properties file without running or deploying associated job."
+      method_option :file_name_prefix, type: :string, aliases: '-p', default: '',
+                    desc: "Add a prefix to job properties filename.This is primarily for use with dryrun"
```
Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: Space needed in "...filename.This..."

Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: Isn't that "for use with dry_run" or "for use with :dry_run"?

File: spec/integration/api/oozie_spec.rb [hunk 17]
```diff
@@ -122,20 +125,14 @@ module Hodor
             }.not_to raise_error
           end
 
-          it "subject should deploy and run the oozie job" do
+          it "should deploy and run the oozie job" do
```
Comment (russellwhitaker) [inline spec/integration/api/oozie_spec.rb]: "Oozie" is capitalized.

File: spec/integration/api/oozie_spec.rb [hunk 18]
```diff
@@ -151,11 +148,51 @@ module Hodor
               oozie.run_job "valid_job"
             }.not_to raise_error
           end
-        end
 
-      end
+          context 'dry_run option is set to true' do
+            it 'subject generates properties file but should not deploy and run the oozie job' do
```
Comment (russellwhitaker) [inline spec/integration/api/oozie_spec.rb]: See previous, apply here and forward.

Comment (russellwhitaker) [inline spec/integration/api/oozie_spec.rb]: I see you're using a mix of single quotes and double quotes around the description strings in your RSpec. I suggest picking one style and sticking with it, unless you need to make specific exceptions for the need to interpolate string values.

File: lib/hodor/api/oozie.rb (hunk 9 above)
Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: Here is body of method Iam going to be checking in as a solution:
` String.new(file_name).insert(1 + file_name.rindex(File::SEPARATOR), prefix)
`

File: lib/tasks/oozie.thor (hunk 13 above)
Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: @dhallman @russellwhitaker  It would be simple to define a constant but a question first.  It looks like unindent is used 14 times in this project.  Values 8 and 10 are passed.  Should these both be 10 or 8 or are there two different formatting options here?

File: lib/hodor/api/oozie.rb (hunk 9 above)
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: I'll ask again: why do you think you need to construct a new string in the body of your method?

Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: Because insert changes the value of the input string.  IE it doesn't change a copy scoped to the block.  At this point we are not using this method in a way where that matters but I prefer not to have that side effect.  An alternative would be to append a ! to method name: append_prefix_to_filename!

Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: Here is Rubydoc on String methods.  
http://ruby-doc.org/core-2.3.0/String.html#method-i-insert

File: lib/tasks/oozie.thor (hunk 13 above)
Comment (dhallman) [inline lib/tasks/oozie.thor]: I put the ```unindent``` in to allow the documentation to align with the source code indent level, while still being left justified when printed to the console. WIthout it, the help text is poorly formatted by thor internals. Most unindent levels are 8 because most thor functions are instance members. The static/class help methods are nested 1 level deeper, hence 2 extra spaces.

Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: How about constants something like CLASS_MEMBER_INDENT=10,  INSTANCE_MEMBER_INDENT=8 in hodor.rb  ?

Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: Let's address this in a separate ticket.   Will add to board.

Comment (russellwhitaker) [inline lib/tasks/oozie.thor]: I agree with @hal-henningsmoen ; I mentioned the one instance I saw, but then later saw a number of different hardcoded indents in the code; we we can deal with this minor carryover technical debt later.

Comment (hal-henningsmoen) [inline lib/tasks/oozie.thor]: Ticket for this is here: https://ulive.leankit.com/Boards/View/229528764/303544652

File: lib/hodor/api/oozie.rb (hunk 8 above)
Comment (russellwhitaker) [inline lib/hodor/api/oozie.rb]: Still not addressed (responding your message about merging).

Comment (hal-henningsmoen) [inline lib/hodor/api/oozie.rb]: Fixing. Big forest tough to see trees.

**Comment by russellwhitaker** [comment]
@hal-henningsmoen I'm confident you've addressed the bulk of my own stated concerns, and thanks for patiently iterating on them. I'm going to assume you're on top of any of the remaining issues, and if @dhallman has had his feedback addressed to his satisfaction, we should be able to merge this with his blessing.

**Comment by dhallman** [comment]
@hal-henningsmoen your changes LGTM. Since @russellwhitaker has already given his LGTM also, I'm going to proceed with the merge.
